scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/model/SchedulePlanGrounder.java

```diff
@@ -0,0 +1,66 @@
+package gov.nasa.jpl.aerie.scheduler.model;
+
+import gov.nasa.jpl.aerie.constraints.model.ActivityInstance;
+import gov.nasa.jpl.aerie.constraints.time.Interval;
+import gov.nasa.jpl.aerie.merlin.driver.ActivityDirective;
+import gov.nasa.jpl.aerie.merlin.driver.ActivityDirectiveId;
+import gov.nasa.jpl.aerie.merlin.driver.StartOffsetReducer;
+import gov.nasa.jpl.aerie.merlin.protocol.types.Duration;
+import org.apache.commons.lang3.tuple.Pair;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class SchedulePlanGrounder {
+  public static Optional<List<ActivityInstance>> groundSchedule(
+      final List<SchedulingActivityDirective> schedulingActivityDirectiveList,
+      final Duration planDuration){
+    final var grounded = new HashMap<ActivityDirectiveId, ActivityInstance>();
+    final var idMap = schedulingActivityDirectiveList.stream().map(a -> Pair.of(new ActivityDirectiveId(a.getId().id()), a)).collect(Collectors.toMap(
+        Pair::getLeft, Pair::getRight));
+    final var converted = schedulingActivityDirectiveList.stream().map(a ->
+                                                                           Pair.of(
+                                                                               new ActivityDirectiveId(a.id().id()),
+                                                                               new ActivityDirective(
+                                                                                a.startOffset(),
+                                                                                a.type().getName(),
+                                                                                a.arguments(),
+                                                                                (a.anchorId() == null) ? null : new ActivityDirectiveId(a.anchorId().id()),
+                                                                                a.anchoredToStart()
+                                                                           ))).collect(Collectors.toMap(Pair::getLeft,
+                                                                                                        Pair::getRight));
+    final var converter = new StartOffsetReducer(planDuration, converted);
+    var computed = converter.compute();
+    computed = StartOffsetReducer.filterOutNegativeStartOffset(computed);
+
+  for(final var entry: computed.entrySet()){
+        Duration offset = Duration.ZERO;
+        final var idActivity = entry.getKey();
+        if(idActivity != null){
+          if(grounded.get(idActivity) == null){
+            return Optional.empty();
+          } else {
+            final var alreadyGroundedAct = grounded.get(idActivity);
+            offset = alreadyGroundedAct.interval.end;
+          }
+        }
+        for(final var activityAndDependents : entry.getValue()) {
```

Comment:
Proposed rename: old name isn't accurate:
```suggestion
        for(final var dependentDirective : entry.getValue()) {
```